### PR TITLE
Add more HF models with bug fixes.

### PR DESCRIPTION
**Pull request by garrett4wade**
1. Add a manual PPO script example.
2. Add deterministic seeding in pytorch dataloader to increase reproducibility.
3. Fix a bug in partitioning micro-batches of paired preference data in the pipe runner. (Related to RW and DPO training).
4. Support several new models: QWen2, Gemma, and GPT-2. `ReaLModel` and its configuration are changed correspondingly to support tied embeddings .
5. Support loading safetensor HF checkpoints.
6. Fix the mapping of contiguous parameters to make the Megatron backend happy.
7. Fix the `finalize_grad` function of the Megatron backend to avoid blocking when training with a partition of all GPUs.
8. Fix gradient all-reduce for LayerNorm and tied embeddings.
9. Add inference tests for new models.
10. Add a parameter reallocation test.

Known issues:
+ The correctness of parameter reallocation is ensured with more assertions at the cost of efficiency. The issue should be resolved with further profiling and implementation.
+ Should check and update the docs.

### CODE REVIEW

File: realhf/impl/model/conversion/hf_registry.py [hunk 1]
```diff
@@ -289,7 +350,9 @@ def save(
...
-        logger.info(
-            f"Saving to HuggingFace Model metadata cost={metadata_t:.2f}s, gather/cpu copy cost={gather_cpu_t:.2f}s, "
-            f"dump cost={dump_t:.2f}s"
-        )
+        if os.getenv("REAL_LOG_SAVE_TIME", None) == "1":
+            logger.info(
+                f"Saving to HuggingFace Model metadata cost={metadata_t:.2f}s, "
+                f"gather/cpu copy cost={gather_cpu_t:.2f}s, "
+                f"dump cost={dump_t:.2f}s"
+            )
```
Comment (nuzant) [inline realhf/impl/model/conversion/hf_registry.py:358]: Is this env var only used here? Maybe set log level.

File: realhf/impl/model/nn/real_llm_api.py [hunk 2]
```diff
@@ -146,8 +185,18 @@ def post_process(self):
         return None
 
     @property
-    def share_embeddings_and_output_weights(self):
-        return self.config.share_embeddings_and_output_weights
+    def shared_embedding_or_output_weights(self) -> bool:
+        return self.config.tied_embedding
+
+    def shared_embedding_or_output_weight(self) -> None | torch.Tensor:
```
Comment (nuzant) [inline realhf/impl/model/nn/real_llm_api.py:191]: Maybe rename one of them? They are too similar.

File: tests/test_distributed_load_hf.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        sd = {}
+        for fn in hf_sd_mapping.values():
+            sd.update(torch.load(path / fn, map_location="cpu"))
+    else:
+        sd = torch.load(path / "pytorch_model.bin", map_location="cpu")
+    return sd
+
+
+def _save_then_load(
+    tmp_path: pathlib.Path,
```
Comment (nuzant) [inline tests/test_distributed_load_hf.py:38]: `pytest` has its own temporary directories and files management, you do not need to make tmp_path of your own. See [this](https://docs.pytest.org/en/stable/how-to/tmp_path.html).

File: tests/test_param_realloc.py [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+    scores = torch.where(prompt_mask, 0, scores)
+
+    loss = ((scores - torch.zeros_like(scores)) ** 2).sum() / (
+        prompt_mask.numel() - prompt_mask.count_nonzero()
+    )
+    return loss, {"loss": loss.clone().detach()}
+
+
+def create_model(
+    tmp_dir: pathlib.Path,
```
Comment (nuzant) [inline tests/test_param_realloc.py:53]: .

File: realhf/api/core/model_api.py [hunk 5]
```diff
@@ -105,28 +105,31 @@ class ReaLModelConfig:
         gradient accumulation in Megatron.
         Currently not supported.
     :type gradient_accumulation_fusion: bool
-    :param share_embeddings_and_output_weights: Whether to share
+    :param tied_embedding: Whether to share
         the embeddings and output weights.
         Currently not supported.
```
Comment (nuzant) [inline realhf/api/core/model_api.py:110]: Update docstring.

File: realhf/impl/model/nn/flatten_param.py [hunk 6]
```diff
@@ -78,22 +70,41 @@ def slice_intervals(
...
+    # offsets = torch.nn.functional.pad(interval_sizes.cumsum(0)[:-1], (1, 0), value=0)
+    # assert tensor.dtype == torch.half
+    # return interval_op_cuda.slice_intervals_cuda_half(
+    #     tensor,
+    #     intervals,
+    #     interval_sizes,
+    #     offsets,
+    #     max_interval_size,
+    #     output_size,
+    # )
```
Comment (nuzant) [inline realhf/impl/model/nn/flatten_param.py:107]: Maybe add a TODO to explain commented code here.

File: realhf/impl/model/nn/flatten_param.py [hunk 7]
```diff
@@ -104,37 +115,81 @@ def set_intervals(
...
+    # offsets = torch.nn.functional.pad(interval_sizes.cumsum(0)[:-1], (1, 0), value=0)
+    # interval_op_cuda.set_intervals_cuda_half(
+    #     src,
+    #     dst,
+    #     intervals,
+    #     interval_sizes,
+    #     offsets,
+    #     max_interval_size,
+    # )
+    # return
```
Comment (nuzant) [inline realhf/impl/model/nn/flatten_param.py:143]: .

**Review by nuzant — COMMENTED** [review]
LGTM, only some minor changes required.

File: realhf/impl/model/nn/real_llm_api.py (hunk 2 above)
Comment (garrett4wade) [reply realhf/impl/model/nn/real_llm_api.py:191]: Reserve `shared_embedding_or_output_weight` in consistent with Megatron-LM. `shared_embedding_or_output_weights` is removed since we do not use Megatron's `finalize_grad` func.

File: realhf/impl/model/conversion/hf_registry.py (hunk 1 above)
Comment (garrett4wade) [reply realhf/impl/model/conversion/hf_registry.py:358]: Changed to `logger.debug`.

File: tests/test_distributed_load_hf.py (hunk 3 above)
Comment (garrett4wade) [reply tests/test_distributed_load_hf.py:38]: fixed

File: tests/test_param_realloc.py (hunk 4 above)
Comment (garrett4wade) [reply tests/test_param_realloc.py:53]: fixed

File: realhf/api/core/model_api.py (hunk 5 above)
Comment (garrett4wade) [reply realhf/api/core/model_api.py]: fixed

File: realhf/impl/model/nn/flatten_param.py (hunk 6 above)
Comment (garrett4wade) [reply realhf/impl/model/nn/flatten_param.py:109]: Added FIXME.

File: realhf/impl/model/nn/flatten_param.py (hunk 7 above)
Comment (garrett4wade) [reply realhf/impl/model/nn/flatten_param.py:148]: Added FIXME.

**Review by nuzant — APPROVED** [review]
